dmenu.el

```diff
@@ -81,18 +81,22 @@ Must be set before initializing Dmenu."
   (let* ((completing-read-fn (if ido-mode
                                  #'ido-completing-read
                                #'completing-read))
-         (execute-file (funcall completing-read-fn dmenu-prompt-string
-                                        (append dmenu--history-list
-                                                (cl-remove-if (lambda (x)
-                                                                (member x dmenu--history-list))
-                                                              dmenu--cache-executable-files))
-                                        nil
-                                        'confirm
-                                        nil
-                                        'dmenu--history-list))
+         (execute-file (funcall completing-read-fn
+                                dmenu-prompt-string
+                                (append dmenu--history-list
+                                        dmenu--cache-executable-files)
+                                nil
+                                'confirm
+                                nil))
          (args (when (= prefix 4)
                  (split-string-and-unquote (read-string "please input the parameters: ")))))
-    (setq dmenu--history-list (cons execute-file (remove execute-file dmenu--history-list)))
+    ;; Add history entries when the user input was a PATH command followed by
+    ;; one or more arguments.
+    (if (let ((cmd+args (split-string execute-file)))
+          (and (> (length cmd+args) 1)
```

Comment:
I think the condition of `(> (length cmd+args) 1)` should be removed. otherwise, the command without arguments would not be cached.

Reply:
Yes, good question.  So, the idea behind this is that the command without arguments is already cached in the list `dmenu--cache-executable-files`.  If we also cache it in `dmenu--history-list` (as the completing read function was doing for us), we then also have to write code to *remove* it from the list of candidates.  That's what (I believe) the `cl-remove-if` call was doing here:
```
(append dmenu--history-list
        (cl-remove-if (lambda (x)
                        (member x dmenu--history-list))
                      dmenu--cache-executable-files))
```
If instead we do NOT admit the command without arguments into `dmenu--history-list` -- which is what this new code is trying to accomplish -- then the `cl-remove-if` call becomes unnecessary.  (See my comment on the next code chunk.)